File: CSX/Callbacks & Higher-order Functions/countBy.js
// COUNTBY

// expected input: array, callbackFn
// expected output: object
// expected result: 

// declare object
// for each element (i) of array
    // if callbackFn(array[i]) does not exist in object
        // assign obj[callbackFn(array[i])] = 1
    // else if callbackFn(array[i]) exists in object
        // increment obj[callbackFn(array[i])]
// return object

function countBy(arr, cb) {
    let obj = {};
    for (let i = 0; i < arr.length; i++) {
        if (Object.keys(obj).includes(cb(arr[i]))) {
            obj[cb(arr[i])]++;
        } else if (!Object.keys(obj).includes(cb(arr[i]))) {
            obj[cb(arr[i])] = 1;
        }
    }
    return obj;
}

function evenOdd(n) {
  if (n % 2 === 0) return 'even';
  else return 'odd';
}
const nums = [1, 2, 3, 4, 5];
console.log(countBy(nums, evenOdd)); // should log: { odd: 3, even: 2 }

File: CSX - Python/Callbacks & Higher-order Functions/count_by.py
def count_by(ls, cb):
    out = {}
    for x in ls:
        res = cb(x)
        if res in out:
            out[res] += 1
        else: out[res] = 1
    return out

def even_odd(n):
    return 'even' if n % 2 == 0 else 'odd'

nums = [1, 2, 3, 4, 5]

print(count_by(nums, even_odd))
